### PR TITLE
GH-97001: Release GIL in termios extension

### PR DESCRIPTION
**Pull request by ronaldoussoren**
Calls to termios APIs might block, therefore give up the GIL while performing these calls.

* Issue: gh-97001

### CODE REVIEW

**Comment by ronaldoussoren** [comment]
IMHO this is a bug fix because you can currently block all Python threads using termios, hence it should be back ported to 3.10 and 3.11.

File: Modules/termios.c [hunk 1]
```diff
@@ -169,7 +174,12 @@ termios_tcsetattr_impl(PyObject *module, int fd, int when, PyObject *term)
     /* Get the old mode, in case there are any hidden fields... */
     termiosmodulestate *state = PyModule_GetState(module);
     struct termios mode;
-    if (tcgetattr(fd, &mode) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+        r = tcgetattr(fd, &mode);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcgetattr(fd, &mode);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 2]
```diff
@@ -82,7 +82,12 @@ termios_tcgetattr_impl(PyObject *module, int fd)
 {
     termiosmodulestate *state = PyModule_GetState(module);
     struct termios mode;
-    if (tcgetattr(fd, &mode) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+       r = tcgetattr(fd, &mode); 
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcgetattr(fd, &mode); 
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 3]
```diff
@@ -211,7 +221,12 @@ termios_tcsetattr_impl(PyObject *module, int fd, int when, PyObject *term)
         return PyErr_SetFromErrno(state->TermiosError);
     if (cfsetospeed(&mode, (speed_t) ospeed) == -1)
         return PyErr_SetFromErrno(state->TermiosError);
-    if (tcsetattr(fd, when, &mode) == -1)
+
+    Py_BEGIN_ALLOW_THREADS
+        r = tcsetattr(fd, when, &mode);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcsetattr(fd, when, &mode);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 4]
```diff
@@ -235,7 +250,13 @@ termios_tcsendbreak_impl(PyObject *module, int fd, int duration)
 /*[clinic end generated code: output=5945f589b5d3ac66 input=dc2f32417691f8ed]*/
 {
     termiosmodulestate *state = PyModule_GetState(module);
-    if (tcsendbreak(fd, duration) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+        r = tcsendbreak(fd, duration);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcsendbreak(fd, duration);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 5]
```diff
@@ -256,7 +277,13 @@ termios_tcdrain_impl(PyObject *module, int fd)
 /*[clinic end generated code: output=5fd86944c6255955 input=c99241b140b32447]*/
 {
     termiosmodulestate *state = PyModule_GetState(module);
-    if (tcdrain(fd) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+        r = tcdrain(fd);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcdrain(fd);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 6]
```diff
@@ -308,7 +341,13 @@ termios_tcflow_impl(PyObject *module, int fd, int action)
 /*[clinic end generated code: output=afd10928e6ea66eb input=c6aff0640b6efd9c]*/
 {
     termiosmodulestate *state = PyModule_GetState(module);
-    if (tcflow(fd, action) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+        r = tcflow(fd, action);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcflow(fd, action);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 7]
```diff
@@ -333,7 +372,13 @@ termios_tcgetwinsize_impl(PyObject *module, int fd)
 #if defined(TIOCGWINSZ)
     termiosmodulestate *state = PyModule_GetState(module);
     struct winsize w;
-    if (ioctl(fd, TIOCGWINSZ, &w) == -1) {
+    int r; 
+
+    Py_BEGIN_ALLOW_THREADS
+        r = ioctl(fd, TIOCGWINSZ, &w);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = ioctl(fd, TIOCGWINSZ, &w);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 8]
```diff
@@ -352,7 +397,12 @@ termios_tcgetwinsize_impl(PyObject *module, int fd)
 #elif defined(TIOCGSIZE)
     termiosmodulestate *state = PyModule_GetState(module);
     struct ttysize s;
-    if (ioctl(fd, TIOCGSIZE, &s) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+        r = ioctl(fd, TIOCGSIZE, &s);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = ioctl(fd, TIOCGSIZE, &s);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 9]
```diff
@@ -282,7 +309,13 @@ termios_tcflush_impl(PyObject *module, int fd, int queue)
 /*[clinic end generated code: output=2424f80312ec2f21 input=0f7d08122ddc07b5]*/
 {
     termiosmodulestate *state = PyModule_GetState(module);
-    if (tcflush(fd, queue) == -1) {
+    int r;
+
+    Py_BEGIN_ALLOW_THREADS
+        r = tcflush(fd, queue);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = tcflush(fd, queue);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 10]
```diff
@@ -433,15 +483,25 @@ termios_tcsetwinsize_impl(PyObject *module, int fd, PyObject *winsz)
         return NULL;
     }
 
-    if (ioctl(fd, TIOCSWINSZ, &w) == -1) {
+    int r;
+    Py_BEGIN_ALLOW_THREADS
+        r = ioctl(fd, TIOCSWINSZ, &w);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = ioctl(fd, TIOCSWINSZ, &w);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 11]
```diff
@@ -433,15 +483,25 @@ termios_tcsetwinsize_impl(PyObject *module, int fd, PyObject *winsz)
...
 
     Py_RETURN_NONE;
 #elif defined(TIOCGSIZE) && defined(TIOCSSIZE)
     struct ttysize s;
+    int r;
     /* Get the old ttysize because it might have more fields. */
-    if (ioctl(fd, TIOCGSIZE, &s) == -1) {
+    Py_BEGIN_ALLOW_THREADS
+        r = ioctl(fd, TIOCGSIZE, &s);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = ioctl(fd, TIOCGSIZE, &s);
    Py_END_ALLOW_THREADS
```

File: Modules/termios.c [hunk 12]
```diff
@@ -453,7 +513,11 @@ termios_tcsetwinsize_impl(PyObject *module, int fd, PyObject *winsz)
         return NULL;
     }
 
-    if (ioctl(fd, TIOCSSIZE, &s) == -1) {
+    Py_BEGIN_ALLOW_THREADS
+        r = ioctl(fd, TIOCSSIZE, &s);
+    Py_END_ALLOW_THREADS
```
Comment (corona10) [inline Modules/termios.c]:
```suggestion
    Py_BEGIN_ALLOW_THREADS
    r = ioctl(fd, TIOCSSIZE, &s);
    Py_END_ALLOW_THREADS
```

**Review by corona10 — COMMENTED** [review]
Overall lgtm,

Super nit: Please unindent between `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS` when considering the style of other code bases.

**Comment by ronaldoussoren** [comment]
> Overall lgtm,
> 
> Super nit: Please unindent between `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS` when considering the style of other code bases.

I'll change that before merging. 

I indent the code between these markers in my own code (including teaching clang-format about this) because this makes it clearer that there's a block of code with slightly different semantics. I prefer my own code style here, but consistency wins.

**Review by corona10 — APPROVED** [review]

**Pull request merged**